backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/interest/InterestComputationService.java

```diff
@@ -0,0 +1,397 @@
+		final BigDecimal interestDistributionPercent = request.getInterestToDistribute().toBigDecimal()
+				.divide(result.getTotalInterestScore(), result.getTotalInterestScore().scale(), RoundingMode.HALF_UP);
+		distributeInterest(request, interestDistributionPercent);
+	}
+
+	private void deletePreviousProgress(@NonNull final PInstanceId logSelectionId)
+	{
+		final ModularLogInterestRepository.LogInterestQuery query = ModularLogInterestRepository.LogInterestQuery.builder()
+				.modularLogSelection(logSelectionId)
+				.build();
+
+		interestRepository.deleteByQuery(query);
+	}
+
+	@NonNull
+	private InitialInterestAllocationResult createInterestRecords(@NonNull final InterestComputationRequest request)
+	{
+		final BigDecimal totalInterestScore = getModularContractIds(request)
+				.stream()
+				.map(contractId -> createInterestRecords(request, contractId))
+				.map(InitialInterestAllocationResult::getTotalInterestScore)
+				.reduce(BigDecimal.ZERO, BigDecimal::add);
+
+		return InitialInterestAllocationResult.of(totalInterestScore);
+	}
+
+	@NonNull
+	private InitialInterestAllocationResult createInterestRecords(
+			@NonNull final InterestComputationRequest request,
+			@NonNull final FlatrateTermId contractId)
+	{
+		final Iterator<ModularContractLogEntry> shippingNotificationIterator = streamShippingNotificationLogEntries(request, contractId).iterator();
+		final Iterator<ModularContractLogEntry> interimInvoiceIterator = streamInterimInvoiceLogEntries(contractId, request).iterator();
+		final ModularContractSettings settings = modularContractService.getModularSettingsForContract(contractId);
+
+		BigDecimal totalInterestScore = BigDecimal.ZERO;
+
+		InvoiceAllocations.AllocationItem currentShippingNotification = null;
+		while (interimInvoiceIterator.hasNext())
+		{
+			final InvoiceAllocations currentInvoiceAllocations = initInvoiceAllocations(request,
+																						settings.getAdditionalInterestDays(),
+																						interimInvoiceIterator.next());
+
+			while (currentInvoiceAllocations.openAmountSignum() > 0 && shippingNotificationIterator.hasNext())
+			{
+				while (shippingNotificationIterator.hasNext() && !currentInvoiceAllocations.canAllocate(currentShippingNotification))
+				{
+					final boolean wasCreatedBeforeInvoice = currentShippingNotification != null
+							&& currentInvoiceAllocations.isInvoiceCreatedAfter(currentShippingNotification);
+
+					if (wasCreatedBeforeInvoice)
+					{
+						saveSubtractValueInterestRecords(request, settings.getBonusInterestRate(), currentShippingNotification);
+					}
+
+					currentShippingNotification = initAllocationItem(request.getInterestToDistribute().getCurrencyId(),
+																	 shippingNotificationIterator.next());
+				}
+
+				if (currentInvoiceAllocations.canAllocate(currentShippingNotification))
+				{
+					currentShippingNotification = currentInvoiceAllocations.allocate(currentShippingNotification);
```

Comment:
pls check, currentShippingNotification might be null

Reply:
it can get pass canAllocate if it's null